src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs

```diff
@@ -8343,6 +8467,146 @@ .locals init (System.Collections.Generic.IEnumerator<int> V_0)
             AssertEx.Equal("System.Int32", op.Info.ElementType.ToTestDisplayString());
         }
 
+        [Fact]
+        public void Foreach_IEnumerableT_LanguageVersion_01()
+        {
+            var src1 = @"
+using System.Collections.Generic;
+
+public ref struct S : IEnumerable<int>
+{
+    public IEnumerator<int> GetEnumerator() => throw null;
+}
+";
+
+            var comp1 = CreateCompilation(src1, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            var src2 = @"
+class C
+{
+    static void Main()
+    {
+        foreach (var i in new S())
+        {
+        }
+    }
+}
+";
+            var comp2 = CreateCompilation(src2, references: [comp1.ToMetadataReference()], targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, parseOptions: TestOptions.Regular12);
+            comp2.VerifyEmitDiagnostics();
+
+            comp2 = CreateCompilation(src2, references: [comp1.ToMetadataReference()], targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, parseOptions: TestOptions.RegularNext);
+            comp2.VerifyEmitDiagnostics();
+
+            comp2 = CreateCompilation(src2, references: [comp1.ToMetadataReference()], targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+            comp2.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void Foreach_IEnumerableT_LanguageVersion_03()
+        {
+            var src1 = @"
+using System.Collections.Generic;
+
+public ref struct S : IEnumerable<int>
+{
+    IEnumerator<int> IEnumerable<int>.GetEnumerator() => throw null;
+    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => throw null;
+}
+";
+
+            var comp1 = CreateCompilation(src1, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            var src2 = @"
+class C
+{
+    static void Main()
+    {
+        foreach (var i in new S())
+        {
+        }
+    }
+}
+";
+            var comp2 = CreateCompilation(src2, references: [comp1.ToMetadataReference()], targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, parseOptions: TestOptions.Regular12);
+            comp2.VerifyEmitDiagnostics();
```

Comment:
>comp2.VerifyEmitDiagnostics(); [](http://example.com/codeflow?start=12&length=30)

Why does this case succeed but the corresponding `-langversion:12` case in `Foreach_IEnumerable_LanguageVersion_03` fails?

Reply:
> Why does this case succeed but the corresponding `-langversion:12` case in `Foreach_IEnumerable_LanguageVersion_03` fails?

Semantic analysis didn't rely and still doesn't rely on conversion classification for generic IEnumerable case. I will check behavior in main if it runs into problems elsewhere and will follow up separately, if needed.